qemu: add emulator allowlist to driver config and add emulator validation

mismithhisler opened the pull request:
### Description
<!-- Please describe why you're making this change and point out any important details the reviewers
should be aware of.-->
These changes add the ability for nomad operators to specify a qemu emulator allowlist for fingerprinting and tasks. If no allowlist is specified, the driver fingerprints all available emulators. Additionally, the emulators are added to a new node attribute `driver.qemu.emulators` which enables using contraints to VM's on nodes with specific emulators, as this functionality is not yet built into the scheduler.

Removed the `fingerprint_emulator` config, which was added in #27128.

### Testing & Reproduction steps
<!--
* In the case of bugs, please describe how to reproduce it.
* If any manual tests were done, document the steps and the conditions to reproduce them.
-->

### Links
<!--
Please include links to GitHub issues, documentation, or similar which is relevant to this PR. If
this is a bug fix, please ensure related issues are linked so they will close when this PR is
merged.
-->

### Contributor Checklist
- [x] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [x] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [x] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the Nomad product documentation, which is stored in the
  [`web-unified-docs` repo](../web-unified-docs/). Refer to the [`web-unified-docs` contributor guide](../web-unified-docs/tree/docs/contribute.md) for docs guidelines.
  Please also consider whether the change requires notes within the [upgrade
  guide](https://developer.hashicorp.com/nomad/docs/upgrade/upgrade-specific). If you would like help with the docs, tag the `nomad-docs` team in this PR.

### Reviewer Checklist
- [ ] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [ ] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository.


<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

- [ ] If a change needs to be reverted, we will roll out an update to the code within 7 days.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

mismithhisler committed:
qemu: add emulator allowlist to driver config and add emulator validation

mismithhisler committed:
add changelog

tgross reviewed (changes requested):
There are some failing tests which I think cover the case where the user hasn't provided an explicit `emulator` argument

mismithhisler committed:
fingerprint all available emulators or those defined in the allowlist

This change removed the fingerprint emulator config parameter and
instead fingerprints all available emulators or those defined in the
allowlist, and adds them to the drivers attribute, allowing placement
contraints if nodes have different emulators installed.

mismithhisler committed:
fix build

mismithhisler commented on drivers/qemu/driver.go:
```diff
@@ -382,6 +396,23 @@ func isAllowedDriveInterface(driveInterface string) bool {
+			return fmt.Errorf("emulator '%s' is not an allowed emulator", emulator)
+		}
+	} else {
+		match, err := regexp.MatchString("qemu-system-*", emulator)
```
I got rid of this in a bit of a refactor, so resolving.

mismithhisler commented on drivers/qemu/driver.go:
```diff
@@ -344,6 +354,45 @@ func (d *Driver) RecoverTask(handle *drivers.TaskHandle) error {
 	return nil
 }
 
+func findEmulators(allowList []string) []string {
```
I will add a godoc to this

mismithhisler committed:
fix validateEmulators test

mismithhisler committed:
add godocs

mismithhisler committed:
use != nil

gulducat reviewed (commented):
lookin pretty good! nice to see qemu getting some love.

couple comments for ya

mismithhisler committed:
update test to skip if aarch64 not detected

mismithhisler committed:
update test

mismithhisler committed:
fix godoc

tgross commented on drivers/qemu/driver.go:
```diff
@@ -344,6 +354,43 @@ func (d *Driver) RecoverTask(handle *drivers.TaskHandle) error {
 
+// findEmulators searches the $PATH for qemu-system binaries until they are found
+// and returns a slice of emulators that comply with the emulators allowlist.
+func findEmulators(allowList []string) []string {
```
Much of this function can be replaced by [`os/exec.LookPath`](https://pkg.go.dev/os/exec#LookPath). We should probably use that so that we're not missing any weird little platform quirks the stdlib is doing for us.

mismithhisler replied:
It's either that or we keep a list of known qemu emulators and just search for each one using exec.LookPath(), although I didn't really want to be maintaining a list like that (although I don't think new architectures are added that often)

tgross replied:
Makes sense. LGTM

mismithhisler commented on drivers/qemu/driver.go:
```diff
@@ -268,8 +275,11 @@ func (d *Driver) buildFingerprint() *drivers.Fingerprint {
+
 	fingerprint.Attributes[driverVersionAttr] = pstructs.NewStringAttribute(currentQemuVersion)
+	fingerprint.Attributes[driverAttr] = pstructs.NewBoolAttribute(true)
+	fingerprint.Attributes[driverEmulatorsAttr] = pstructs.NewStringAttribute(strings.Join(emulators, ","))
```
Looking at this now, I think we're protected when we validate the emulator is in the allowlist.

mismithhisler committed:
small refactor

tgross reviewed (approved):
LGTM!
